File: Jessica/108.Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//每次都是取中间
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return convertToBst(nums, 0, nums.size()-1);
        
    }
    
TreeNode* convertToBst(vector<int>& nums, int begin, int end)
{
    if (begin > end)
        return NULL;
    
    if (begin == end)
    {
        return new TreeNode(nums[begin]);
    }
    int mid = (begin + end)/2;
    TreeNode* root = new TreeNode(nums[mid]);
    root->left = convertToBst(nums,begin, mid-1 );
    root->right = convertToBst(nums,mid+1, end);
    return root;
        
    }
    
};

File: Jessica/108.Convert Sorted Array to Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#list中间数做root,左右两边递归
class Solution:
    def Recursion(self, nums: List[int], start: int, end: int) -> TreeNode:
        if start > end:
            return None
        elif start == end:
            return TreeNode(nums[start])
        else:
            mid = (start + end) // 2
            node = TreeNode(nums[mid])
            node.left = self.Recursion(nums, start, mid - 1)
            node.right = self.Recursion(nums, mid + 1, end)
            return node
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.Recursion(nums, 0, len(nums) - 1)
    
    
